File: archive/canvas-api/src/main/java/com/instructure/canvasapi/api/StreamAPI.java
/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 *     Licensed under the Apache License, Version 2.0 (the "License");
 *     you may not use this file except in compliance with the License.
 *     You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *     Unless required by applicable law or agreed to in writing, software
 *     distributed under the License is distributed on an "AS IS" BASIS,
 *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *     See the License for the specific language governing permissions and
 *     limitations under the License.
 *
 */

package com.instructure.canvasapi.api;

import android.content.Context;

import com.instructure.canvasapi.model.CanvasContext;
import com.instructure.canvasapi.model.HiddenStreamItem;
import com.instructure.canvasapi.model.StreamItem;
import com.instructure.canvasapi.utilities.APIHelpers;
import com.instructure.canvasapi.utilities.CanvasCallback;
import com.instructure.canvasapi.utilities.CanvasRestAdapter;

import retrofit.Callback;
import retrofit.RestAdapter;
import retrofit.http.DELETE;
import retrofit.http.Path;
import retrofit.http.GET;
import retrofit.http.Query;


public class StreamAPI extends BuildInterfaceAPI {

    interface StreamInterface {
        @GET("/users/self/activity_stream")
        void getUserStream(Callback<StreamItem[]> callback);

        @DELETE("/users/self/activity_stream/{streamID}")
        void hideStreamItem(@Path("streamID")long streamID, Callback<HiddenStreamItem> callback);

        @GET("/{context_id}/activity_stream")
        void getContextStream(@Path("context_id") long context_id, Callback<StreamItem[]> callback);

        @GET("/{next}")
        void getNextPageStream(@Path(value = "next", encode = false) String nextURL, Callback<StreamItem[]> callback);

        /////////////////////////////////////////////////////////////////////////////
        // Synchronous
        /////////////////////////////////////////////////////////////////////////////

        @GET("/users/self/activity_stream")
        StreamItem[] getUserStreamSynchronous(@Query("per_page") int number);
    }


    /////////////////////////////////////////////////////////////////////////
    // API Calls
    /////////////////////////////////////////////////////////////////////////

    public static void getFirstPageUserStream(CanvasCallback<StreamItem[]> callback) {
        if (APIHelpers.paramIsNull(callback)) { return; }

        buildCacheInterface(StreamInterface.class, callback, null).getUserStream(callback);
        buildInterface(StreamInterface.class, callback, null).getUserStream(callback);
    }

    public static void getFirstPageCourseStream(CanvasContext canvasContext, CanvasCallback<StreamItem[]> callback) {
        if (APIHelpers.paramIsNull(callback ,canvasContext)) { return; }

        buildCacheInterface(StreamInterface.class, callback, canvasContext).getContextStream(canvasContext.getId(), callback);
        buildInterface(StreamInterface.class, callback, canvasContext).getContextStream(canvasContext.getId(), callback);
    }

    public static void getNextPageStream(String nextUrl, CanvasCallback<StreamItem[]> callback) {
        if (APIHelpers.paramIsNull(callback, nextUrl)) { return; }

        callback.setIsNextPage(true);

        buildCacheInterface(StreamInterface.class, callback, false).getNextPageStream(nextUrl, callback);
        buildInterface(StreamInterface.class, callback, false).getNextPageStream(nextUrl, callback);
    }

    public static void hideStreamItem(long streamId, final CanvasCallback<HiddenStreamItem> callback) {
        if (APIHelpers.paramIsNull(callback)) { return; }
        buildInterface(StreamInterface.class, callback, null).hideStreamItem(streamId,callback);
    }

    /////////////////////////////////////////////////////////////////////////////
    // Synchronous
    //
    // If Retrofit is unable to parse (no network for example) Synchronous calls
    // will throw a nullPointer exception. All synchronous calls need to be in a
    // try catch block.
    /////////////////////////////////////////////////////////////////////////////

    public static StreamItem[] getUserStreamSynchronous(Context context, int numberToReturn) {
        try {
            RestAdapter restAdapter = CanvasRestAdapter.buildAdapter(context);

            return restAdapter.create(StreamInterface.class).getUserStreamSynchronous(numberToReturn);
        } catch (Exception E){
            return null;
        }
    }
}

File: libs/canvas-api-2/src/main/java/com/instructure/canvasapi2/apis/StreamAPI.kt
/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 *     Licensed under the Apache License, Version 2.0 (the "License");
 *     you may not use this file except in compliance with the License.
 *     You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *     Unless required by applicable law or agreed to in writing, software
 *     distributed under the License is distributed on an "AS IS" BASIS,
 *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *     See the License for the specific language governing permissions and
 *     limitations under the License.
 */
package com.instructure.canvasapi2.apis

import com.instructure.canvasapi2.StatusCallback
import com.instructure.canvasapi2.builders.RestBuilder
import com.instructure.canvasapi2.builders.RestParams
import com.instructure.canvasapi2.models.CanvasContext
import com.instructure.canvasapi2.models.HiddenStreamItem
import com.instructure.canvasapi2.models.StreamItem
import com.instructure.canvasapi2.utils.DataResult
import retrofit2.Call
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.Query
import retrofit2.http.Tag
import retrofit2.http.Url

object StreamAPI {

    interface StreamInterface {
        @GET("users/self/activity_stream?only_active_courses=true")
        fun getUserStream(): Call<List<StreamItem>>

        @GET("users/self/activity_stream?only_active_courses=true")
        suspend fun getUserStream(@Tag params: RestParams): DataResult<List<StreamItem>>

        @GET("{contextId}/activity_stream?only_active_courses=true")
        fun getContextStream(@Path("contextId") contextId: Long): Call<List<StreamItem>>

        @GET
        fun getNextPageStream(@Url nextURL: String): Call<List<StreamItem>>

        @DELETE("users/self/activity_stream/{streamId}")
        fun hideStreamItem(@Path("streamId") streamId: Long): Call<HiddenStreamItem>

        @GET("users/self/activity_stream?only_active_courses=true")
        fun getUserStreamCustomCount(@Query("per_page") number: Int): Call<List<StreamItem>>
    }

    fun getUserStream(adapter: RestBuilder, params: RestParams, callback: StatusCallback<List<StreamItem>>) {
        if (StatusCallback.isFirstPage(callback.linkHeaders)) {
            callback.addCall(adapter.build(StreamInterface::class.java, params).getUserStream()).enqueue(callback)
        } else if (callback.linkHeaders != null && StatusCallback.moreCallsExist(callback.linkHeaders)) {
            callback.addCall(adapter.build(StreamInterface::class.java, params).getNextPageStream(callback.linkHeaders!!.nextUrl!!)).enqueue(callback)
        }
    }

    fun getCourseStream(canvasContext: CanvasContext, adapter: RestBuilder, params: RestParams, callback: StatusCallback<List<StreamItem>>) {
        if (StatusCallback.isFirstPage(callback.linkHeaders)) {
            callback.addCall(adapter.build(StreamInterface::class.java, params).getContextStream(canvasContext.id)).enqueue(callback)
        } else if (callback.linkHeaders != null && StatusCallback.moreCallsExist(callback.linkHeaders)) {
            callback.addCall(adapter.build(StreamInterface::class.java, params).getNextPageStream(callback.linkHeaders!!.nextUrl!!)).enqueue(callback)
        }
    }

    fun hideStreamItem(streamId: Long, adapter: RestBuilder, params: RestParams, callback: StatusCallback<HiddenStreamItem>) {
        callback.addCall(adapter.build(StreamInterface::class.java, params).hideStreamItem(streamId)).enqueue(callback)
    }

    fun getUserStreamSynchronous(numberToReturn: Int, adapter: RestBuilder, params: RestParams): List<StreamItem>? {
        return try {
            adapter.build(StreamInterface::class.java, params).getUserStreamCustomCount(numberToReturn).execute().body()
        } catch (E: Exception) {
            null
        }
    }
}
